File: backend/app/schemas/scenario.py
"""Pydantic schemas for scenario mode API."""

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, field_validator

from app.schemas.anomaly import VALID_ANOMALY_TYPES


class ScenarioStepCreate(BaseModel):
    """Schema for creating a single scenario step."""

    register_name: str
    anomaly_type: str
    anomaly_params: dict[str, Any] = {}
    trigger_at_seconds: int
    duration_seconds: int
    sort_order: int = 0

    @field_validator("anomaly_type")
    @classmethod
    def validate_anomaly_type(cls, v: str) -> str:
        if v not in VALID_ANOMALY_TYPES:
            raise ValueError(f"anomaly_type must be one of {VALID_ANOMALY_TYPES}")
        return v

    @field_validator("trigger_at_seconds")
    @classmethod
    def validate_trigger_at_seconds(cls, v: int) -> int:
        if v < 0:
            raise ValueError("trigger_at_seconds must be >= 0")
        return v

    @field_validator("duration_seconds")
    @classmethod
    def validate_duration_seconds(cls, v: int) -> int:
        if v <= 0:
            raise ValueError("duration_seconds must be > 0")
        return v


class ScenarioCreate(BaseModel):
    """Schema for creating a scenario."""

    template_id: UUID
    name: str
    description: str | None = None
    steps: list[ScenarioStepCreate]


class ScenarioUpdate(BaseModel):
    """Schema for updating a scenario."""

    name: str
    description: str | None = None
    steps: list[ScenarioStepCreate]


class ScenarioStepResponse(BaseModel):
    """Response schema for a scenario step."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    register_name: str
    anomaly_type: str
    anomaly_params: dict[str, Any]
    trigger_at_seconds: int
    duration_seconds: int
    sort_order: int


class ScenarioSummary(BaseModel):
    """Summary response for a scenario (list view)."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    template_id: UUID
    template_name: str
    name: str
    description: str | None
    is_builtin: bool
    total_duration_seconds: int
    created_at: datetime
    updated_at: datetime


class ScenarioDetail(ScenarioSummary):
    """Detail response for a scenario (includes steps)."""

    steps: list[ScenarioStepResponse]


class ScenarioExport(BaseModel):
    """Schema for exporting a scenario as portable JSON."""

    name: str
    description: str | None
    template_name: str
    steps: list[ScenarioStepCreate]


class ActiveStepStatus(BaseModel):
    """Status of a currently active step during execution."""

    register_name: str
    anomaly_type: str
    remaining_seconds: int


class ScenarioExecutionStatus(BaseModel):
    """Real-time execution status of a running scenario."""

    scenario_id: UUID
    scenario_name: str
    status: str
    elapsed_seconds: int
    total_duration_seconds: int
    active_steps: list[ActiveStepStatus]

File: frontend/src/types/scenario.ts
export interface ScenarioStepCreate {
  register_name: string;
  anomaly_type: string;
  anomaly_params: Record<string, number | string | boolean>;
  trigger_at_seconds: number;
  duration_seconds: number;
  sort_order: number;
}

export interface ScenarioStepResponse extends ScenarioStepCreate {
  id: string;
}

export interface ScenarioSummary {
  id: string;
  template_id: string;
  template_name: string;
  name: string;
  description: string | null;
  is_builtin: boolean;
  total_duration_seconds: number;
  created_at: string;
  updated_at: string;
}

export interface ScenarioDetail extends ScenarioSummary {
  steps: ScenarioStepResponse[];
}

export interface ScenarioCreate {
  template_id: string;
  name: string;
  description?: string | null;
  steps: ScenarioStepCreate[];
}

export interface ScenarioUpdate {
  name: string;
  description?: string | null;
  steps: ScenarioStepCreate[];
}

export interface ScenarioExport {
  name: string;
  description: string | null;
  template_name: string;
  steps: ScenarioStepCreate[];
}

export interface ActiveStepStatus {
  register_name: string;
  anomaly_type: string;
  remaining_seconds: number;
}

export interface ScenarioExecutionStatus {
  scenario_id: string;
  scenario_name: string;
  status: "running" | "completed";
  elapsed_seconds: number;
  total_duration_seconds: number;
  active_steps: ActiveStepStatus[];
}
